ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/upnp/UpnpNatManager.java

```diff
@@ -0,0 +1,487 @@
+                         * signature, we must too when we override it.
+                         */
+                        @Override
+                        @SuppressWarnings("rawtypes")
+                        public void failure(
+                            final ActionInvocation invocation,
+                            final UpnpResponse operation,
+                            final String msg) {
+                          upnpQueryFuture.completeExceptionally(new Exception(msg));
+                        }
+                      };
+                  upnpService.getControlPoint().execute(callback);
+
+                  return upnpQueryFuture;
+                });
+  }
+
+  /**
+   * Sends a UPnP request to the discovered IGD to request status info.
+   *
+   * @return A CompletableFuture that can be used to query the result (or error).
+   */
+  public CompletableFuture<Connection.StatusInfo> queryStatusInfo() {
+
+    CompletableFuture<Connection.StatusInfo> upnpQueryFuture = new CompletableFuture<>();
+
+    return discoverService(SERVICE_TYPE_WAN_IP_CONNECTION)
+        .thenCompose(
+            service -> {
+              GetStatusInfo callback =
+                  new GetStatusInfo(service) {
+                    @Override
+                    public void success(final Connection.StatusInfo statusInfo) {
+                      upnpQueryFuture.complete(statusInfo);
+                    }
+
+                    /**
+                     * Because the underlying jupnp library omits generics info in this method
+                     * signature, we must too when we override it.
+                     */
+                    @Override
+                    @SuppressWarnings("rawtypes")
+                    public void failure(
+                        final ActionInvocation invocation,
+                        final UpnpResponse operation,
+                        final String msg) {
+                      upnpQueryFuture.completeExceptionally(new Exception(msg));
+                    }
+                  };
+              upnpService.getControlPoint().execute(callback);
+
+              return upnpQueryFuture;
+            });
+  }
+
+  /**
+   * Convenience function to call {@link #requestPortForward(PortMapping)} with the following
+   * defaults:
+   *
+   * <p>enabled: true leaseDurationSeconds: 0 (indefinite) remoteHost: null internalClient: the
+   * local address used to discover gateway
+   *
+   * <p>In addition, port is used for both internal and external port values.
+   *
+   * @param port is the port to be used for both internal and external port values
+   * @param protocol is either "udp" or "tcp"
+   * @param description is a free-form description, often displayed in router UIs
+   * @return A CompletableFuture which will provide the results of the request
+   */
+  public CompletableFuture<Void> requestPortForward(
+      final int port, final String protocol, final String description) {
```

Comment:
jupnp provides an enum internally:

``` java
public class PortMapping {

    public enum Protocol {
        UDP,
        TCP
    }
    // ...
}
```

However, I wanted to avoid exposing PortMapping, so I also wanted to avoid exposing PortMapping.Protocol. On the other hand, creating a new enum just to avoid exposing this one seemed ridiculous, so I settled on leaving it as a String.

I'll revisit this -- I agree that a String is not ideal here.